File: tax-mcp/tools/receipts.py
import json
from typing import Optional
from google.cloud.firestore import Client
from mcp.server.fastmcp import FastMCP


def register_receipt_tools(mcp: FastMCP, db: Client, user_id: str):

    def _col(tax_year_id: str):
        return (
            db.collection("users").document(user_id)
            .collection("taxYears").document(tax_year_id)
            .collection("receipts")
        )

    @mcp.tool()
    def list_receipts(tax_year_id: str, status: Optional[str] = None) -> str:
        """List receipts. status: PENDING | PROCESSING | EXTRACTED | VERIFIED | FAILED"""
        q = _col(tax_year_id).order_by("createdAt", direction="DESCENDING")
        if status:
            q = q.where("status", "==", status)
        return json.dumps([{"id": d.id, **d.to_dict()} for d in q.stream()], default=str)

    @mcp.tool()
    def get_receipt(tax_year_id: str, receipt_id: str) -> str:
        """Get a single receipt by ID."""
        doc = _col(tax_year_id).document(receipt_id).get()
        if not doc.exists:
            raise ValueError(f"Receipt {receipt_id} not found")
        return json.dumps({"id": doc.id, **doc.to_dict()}, default=str)

File: libs/tax-mcp/src/lib/tools/receipts.ts
import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { z } from 'zod';
import { receiptsCol, receiptDoc } from '@can-tax-pro/db';
import { ReceiptStatus } from '@can-tax-pro/types';

export function registerReceiptTools(server: McpServer, userId: string) {
  server.tool(
    'list_receipts',
    'List receipts for a tax year, optionally filtered by status',
    {
      taxYearId: z.string(),
      status: z.nativeEnum(ReceiptStatus).optional(),
    },
    async ({ taxYearId, status }) => {
      let q = receiptsCol(userId, taxYearId).orderBy('createdAt', 'desc') as FirebaseFirestore.Query;
      if (status) q = q.where('status', '==', status);
      const snap = await q.get();
      const receipts = snap.docs.map((d) => ({ id: d.id, ...d.data() }));
      return { content: [{ type: 'text', text: JSON.stringify(receipts, null, 2) }] };
    }
  );

  server.tool(
    'get_receipt',
    'Get a single receipt by ID',
    { taxYearId: z.string(), receiptId: z.string() },
    async ({ taxYearId, receiptId }) => {
      const doc = await receiptDoc(userId, taxYearId, receiptId).get();
      if (!doc.exists) throw new Error(`Receipt ${receiptId} not found`);
      return { content: [{ type: 'text', text: JSON.stringify({ id: doc.id, ...doc.data() }) }] };
    }
  );
}
